railgun/hack/generators/manifests/main.go

```diff
@@ -0,0 +1,124 @@
+// and if you try to apply a v1 CRD that does NOT have this field set where a
+// v1beta1 version of the CRD already exists on the cluster this can result in
+// and error and fail to upgrade the CRD.
+//
+// This program accounts for this by applying "preserveUnknownFields: false" to
+// any v1 CustomResourceDefinition to work around the upstream backwards
+// incompatibility issue automatically.
+//
+// NOTE: When this was first written using controller-gen configuration options
+//       and kubebuilder flags on the API types was attempted but none of these
+//       things could produce the right result.
+//
+//       The controller-gen CLI can't really do it (easily) because the default
+//       behavior of the marshaller in sigs.k8s.io/yaml is to not emit optional
+//       boolean fields to YAML if they're set to false. 😑
+//
+//       It's quite possible there's a better way to do this and if you find one
+//       we encourage you to put in a PR and replace this.
+//
+// -----------------------------------------------------------------------------
+
+func main() {
+	// handle arguments
+	flag.StringVar(&directory, "directory", ".", "which directory to find CRDs in")
+	flag.Parse()
+
+	// find all the YAML CRDs in the provided directory and patch them
+	if err := filepath.Walk(directory, processFile); err != nil {
+		fmt.Fprintf(os.Stderr, "Error processing files in %s: %s\n", directory, err)
+		os.Exit(1)
+	}
+}
+
+func processFile(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+
+	if info.IsDir() {
+		return nil
+	}
+
+	// filter out any files that don't identify themselves as YAML files via file extensions
+	if !strings.HasSuffix(info.Name(), ".yaml") && !strings.HasSuffix(info.Name(), ".yml") {
+		return nil
+	}
+
+	// read in the file contents
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+
+	// filter out any YAML files that don't contain a v1.CustomResourceDefinition
+	if !bytes.Contains(b, []byte(`kind: CustomResourceDefinition`)) || !bytes.Contains(b, []byte(`apiVersion: apiextensions.k8s.io/v1`)) {
+		return nil
+	}
+
+	// don't allow files that contain multiple objects (controller-gen wont be configured to emit these anyway)
+	yamlCRDs := bytes.Split(b, []byte(`\n---\n`))
+	if len(yamlCRDs) > 1 {
+		return fmt.Errorf("file %s contained multiple objects (%d) which this script doesn't yet support", path, len(yamlCRDs))
+	}
+
+	// generate a patch command to ensure "preserveUnknownFields" is set to false
+	stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)
+	patchCMD := exec.Command("kubectl", "patch", "--type", "merge", "--local", "-o", "yaml", "-f", path, "--patch-file", patchfile)
+	patchCMD.Stdout = stdout
+	patchCMD.Stderr = stderr
+
+	// run the patcher and make sure there are no errors
+	if err := patchCMD.Run(); err != nil {
+		return fmt.Errorf("failed to patch crd in %s: %w; STDOUT=(%s) STDERR=(%s)",
```

Comment:
migration, upgrade makes more sense ?

Reply:
In this context the topical verb is "patch" because all this file does is use `kubectl patch` at the end of the day, so I don't think `migration` or `upgrade` makes any sense because those aren't things we're actually doing as a part of this script.

Reply:
sorry, but which file/line is executing "kubectl patch " ?

Reply:
line `113` is executing `kubectl patch` (line `108` defines it)

Reply:
Line 108 does patch 
Its very expensive executing "kubectl patch" within a golang module and it is not a test case, could you please change to use client-go library  instead ?

Reply:
@ccfishk, see https://github.com/Kong/kubernetes-ingress-controller/pull/1559#discussion_r671287102 which is the context behind the decision to use `kubectl patch`.

Note that we perform `kubectl patch` locally, without interaction with any kubernetes API server. Our intention is to only edit a YAML file on the disk by adding a field to it. The PR linked above is the reason why we opted against deserializing and serializing the CRD YAML and editing it programmatically in Go.

My suggestion is to keep `kubectl patch` here as-is, so that we don't need to maintain the patching logic.

Reply:
@shaneutt  and Michal, lets leave as it is.
Lets get this merged asap.